File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    //Prompt user for a credit card number and check length for validity.

    long x = get_long("Credit Card Number: \n");
    if (x < 1000000000000 || x > 9999999999999999)
    {
        printf("INVALID\n");
        return 0;
    }
    //Now we have a 13, 15 or 16 digit credit card number

    //Isolate and double every other digit, starting from the second to last
    int y = 1, z = 1, pot = 0;
    for (int i = 0; i < 8; i++)
    {
        long a = x % (long)pow(10,2*y);
        long b = (long)pow(10,(2 * y - 1));
        long c = (a - (a % b)) / b;
        long d = 2 * c;
        if (d >= 10)
        {
            int r = d % 10;
            int o = (d - r) / 10;
            pot = pot + r + o;
        }
        else
        {
           pot = pot + d;
        }
        printf("cc digit: %li | pot: %i\n", c, pot);
        y++;
     }

    //Isolate every other digit starting with the last one, and add it to the pot

    for (int i = 0; i < 8; i++)
    {
        if (z == 1)
        {
            long d = x % 10;
            pot = pot + d;
            printf("cc digit: %li | pot: %i\n", d, pot);
        }
        else
        {
            long a = x % (long)pow(10,2 * z - 1);
            long b = (long)pow(10,(2 * z - 2));
            long d = (a - (a % b)) / b;
            pot = pot + d;
            printf("cc digit: %li | pot: %i\n", d, pot);
        }
        z++;
    }

        
    //Check sum % 10
    if (pot % 10 == 0)
    {
        int first = 0;
        //isolate the first two digits to check card type. Use int datatype to cut off decimals
        if (x > 999999999999 && x < 10000000000000) //13 digit
        {
            first = x / pow(10,11);
        }
        else if (x > 99999999999999 && x < 1000000000000000) //15 digit
        {
            first = x / pow(10,13);
        }
        else if (x > 999999999999999 && x < 10000000000000000) //16 digit
        {
            first = x / pow(10,14);
        }
        else
        {
            printf("INVALID\n"); //there are no 14 digit cards, exit the program if one got through the test.
             return 0;
        }

        if (first > 39 && first < 50)
        {
            printf("VISA\n");
             return 0;
        }
        else if (first == 34 || first == 37)
        {
            printf("AMEX\n");
             return 0;
        }
        else if (first > 50 && first < 56)
        {
            printf("MASTERCARD\n");
             return 0;
        }
        else
        {
            printf("INVALID\n");
             return 0;
        }
    }
    else
    {
        printf("INVALID\n");
         return 0;
    }
}

File: credit.py
# this program checks a credit card against Luhn's Algorithm


from cs50 import get_float, get_int, get_string


def main():

    ccNum = get_int("please enter a credit card number for testing: ")
    length = len(str(ccNum))
    if length < 13 or length > 17:
        print("INVALID")
        return 1

    if checksum(ccNum):
        # valid credit card number, now check issuer
        # All American Express numbers start with 34 or 37
        if int(str(ccNum)[0]) == 3 and (int(str(ccNum)[1]) == 4 or int(str(ccNum)[1]) == 7):
            print("AMEX")
        # most MasterCard numbers start with 51, 52, 53, 54, or 55
        elif int(str(ccNum)[0]) == 5 and (int(str(ccNum)[1]) > 0 and int(str(ccNum)[1]) < 6):
            print("MASTERCARD")
        # and all Visa numbers start with 4.
        elif int(str(ccNum)[0]) == 4:
            print("VISA")
        else:
            print("INVALID")
            exit(2)


def checksum(ccNum):
    # Luhn's Algorithm:
    # multiply every other digit, starting with the last
    index = -1
    bucket = 0
    for i in range(len(str(ccNum))):
        if i % 2 == 0:
            # add the digits from ccNum what were not doubled (odd reverse index)
            bucket += int(str(ccNum)[index])

        else:
            # double the digit (odd reverse index)
            temp = int(str(ccNum)[index]) * 2
            if temp > 9:
                bucket += (temp - 9)
            else:
                bucket += temp

        index -= 1

    # if total % 10 = 0, the cc number is valid
    if bucket % 10 == 0:
        return True
    else:
        print("INVALID")
        exit(1)


if __name__ == "__main__":
    main()